src/compiler/utilities.ts

```diff
@@ -2056,19 +2056,51 @@ namespace ts {
             }
             return AssignmentDeclarationKind.ObjectDefinePropertyValue;
         }
-        if (expr.operatorToken.kind !== SyntaxKind.EqualsToken ||
-            !isPropertyAccessExpression(expr.left)) {
+        if (expr.operatorToken.kind !== SyntaxKind.EqualsToken) {
             return AssignmentDeclarationKind.None;
         }
-        const lhs = expr.left;
-        if (isEntityNameExpression(lhs.expression) && lhs.name.escapedText === "prototype" && isObjectLiteralExpression(getInitializerOfBinaryExpression(expr))) {
-                // F.prototype = { ... }
-                return AssignmentDeclarationKind.Prototype;
+        if (isPropertyAccessExpression(expr.left) || isElementAccessExpression(expr.left)) {
+            const lhs = expr.left;
+            if (isEntityNameExpression(lhs.expression) && getElementOrPropertyAccessName(lhs) === "prototype" && isObjectLiteralExpression(getInitializerOfBinaryExpression(expr))) {
+                    // F.prototype = { ... }
+                    return AssignmentDeclarationKind.Prototype;
+            }
+            return getAssignmentDeclarationPropertyAccessKind(lhs);
+        }
+        return AssignmentDeclarationKind.None;
+    }
+
+    // TODO: Signed numeric literals?

```

Comment:
Decide whether they're in or out and either remove the TODO or replace it with a warning that they are(n't) supported.

Reply:
They should proooobably be in, but some checks are going to get way more complicated to support them.

Reply:
> replace it with a warning that they are(n't) supported.

That would be a strict degradation from today's behavior where `A[+0]` is simply an access on `A`'s `0` - no we can't warn here; but I do think recognizing more than this is just an incremental enhancement.

Reply:
For reference, if you write
```js
const x = {
    [+(0)]: 12
}

const n = x[1]
```
`n` is not an error, but if you write
```js
const x = {
    [+0]: 12
}

const n = x[1]
```
it is, the only difference being a runtime non-affecting set of parenthesis.

Reply:
yeah, a user warning isn't necessary, but a jsdoc warning instead of a TODO in the source is good enough.